placement/algo/algo.go

```diff
@@ -0,0 +1,470 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package algo
+
+import (
+	"container/heap"
+	"errors"
+	"math"
+
+	"github.com/m3db/m3cluster/placement"
+)
+
+var (
+	errNotEnoughRacks          = errors.New("not enough racks to take shards, please make sure RF <= number of racks")
+	errHostAbsent              = errors.New("could not remove or replace a host that does not exist")
+	errCouldNotReachTargetLoad = errors.New("new host could not reach target load")
+)
+
+type rackAwarePlacementAlgorithm struct {
+}
+
+// NewRackAwarePlacementAlgorithm returns a rack aware placement algorithm
+func NewRackAwarePlacementAlgorithm() placement.Algorithm {
+	return rackAwarePlacementAlgorithm{}
+}
+
+func (a rackAwarePlacementAlgorithm) BuildInitialPlacement(hosts []placement.Host, shards []uint32) (placement.Snapshot, error) {
+	ph := newInitPlacementHelper(hosts, shards)
+
+	if err := ph.placeShards(ph.uniqueShards, nil); err != nil {
+		return nil, err
+	}
+	return ph.generatePlacement(), nil
+}
+
+func (a rackAwarePlacementAlgorithm) AddReplica(ps placement.Snapshot) (placement.Snapshot, error) {
+	ph := newReplicaPlacementHelper(ps, ps.Replicas()+1)
+	if err := ph.placeShards(ph.uniqueShards, nil); err != nil {
+		return nil, err
+	}
+	return ph.generatePlacement(), nil
+}
+
+func (a rackAwarePlacementAlgorithm) RemoveHost(ps placement.Snapshot, leavingHost placement.Host) (placement.Snapshot, error) {
+	if a.isHostAbsent(ps, leavingHost) {
+		return nil, errHostAbsent
+	}
+	ph, leavingHostShards := newRemoveHostPlacementHelper(ps, leavingHost)
+	// place the shards from the leaving host to the rest of the cluster
+	if err := ph.placeShards(leavingHostShards.Shards(), leavingHostShards); err != nil {
+		return nil, err
+	}
+	return ph.generatePlacement(), nil
+}
+
+func (a rackAwarePlacementAlgorithm) AddHost(ps placement.Snapshot, addingHost placement.Host) (placement.Snapshot, error) {
+	ph, addingHostShards := newAddHostPlacementHelper(ps, addingHost)
+	targetLoad := ph.hostHeap.getTargetLoadForHost(addingHostShards.hostAddress())
+	// try to steal shards from the most loaded hosts until the adding host reaches target load
+outer:
+	for len(addingHostShards.shardsSet) < targetLoad {
+		if ph.hostHeap.Len() == 0 {
+			return nil, errCouldNotReachTargetLoad
+		}
+		tryHost := heap.Pop(ph.hostHeap).(*hostShards)
+		for shard := range tryHost.shardsSet {
+			if ph.canAssignHost(shard, tryHost, addingHostShards) {
+				ph.assignShardToHost(shard, addingHostShards)
```

Comment:
Perhaps we can have a `moveShard(shard, fromHost, toHost)` and call it here instead?